File: server/tests/unit/exerciseTrackingTypes.test.ts
import { prescribedFieldsFor, loggedFieldsFor, targetOnlyFieldsFor } from '../../src/config/exerciseTrackingTypes';

// Regression test: loggedFieldsFor() used to filter its result through
// fieldOrder (the *prescribed* canonical order), which never included
// "weight" since training_sets has no weight column. That silently dropped
// weight from every logged view (session logging inputs, workout history)
// for every Sets & Reps exercise, regardless of the coach's tracked_metrics.
describe('loggedFieldsFor', () => {
    test('sets_reps always includes weight and reps, regardless of tracked_metrics', () => {
        expect(loggedFieldsFor({ tracking_type: 'sets_reps', tracked_metrics: [] })).toEqual(['weight', 'reps']);
        expect(loggedFieldsFor({ tracking_type: 'sets_reps', tracked_metrics: ['tempo', 'rir', 'rpe'] })).toEqual(['weight', 'reps']);
    });

    test('time_based logs only the coach-selected metrics, in canonical order', () => {
        expect(loggedFieldsFor({ tracking_type: 'time_based', tracked_metrics: [] })).toEqual([]);
        expect(loggedFieldsFor({ tracking_type: 'time_based', tracked_metrics: ['speed_kmh', 'duration_seconds'] }))
            .toEqual(['duration_seconds', 'speed_kmh']); // order follows fieldOrder, not input order
    });

    test('falls back to sets_reps for an ad-hoc exercise with no tracking_type', () => {
        expect(loggedFieldsFor(null)).toEqual(['weight', 'reps']);
        expect(loggedFieldsFor(undefined)).toEqual(['weight', 'reps']);
    });
});

describe('prescribedFieldsFor', () => {
    test('sets_reps never includes weight (training_sets has no weight column)', () => {
        const fields = prescribedFieldsFor({ tracking_type: 'sets_reps', tracked_metrics: ['tempo', 'rir', 'rpe'] });
        expect(fields).not.toContain('weight');
        expect(fields).toEqual(['reps', 'rest_seconds', 'tempo', 'rir', 'rpe']);
    });

    test('time_based always includes rest_seconds plus selected metrics', () => {
        expect(prescribedFieldsFor({ tracking_type: 'time_based', tracked_metrics: ['distance_km'] }))
            .toEqual(['distance_km', 'rest_seconds']);
    });
});

describe('targetOnlyFieldsFor', () => {
    test('sets_reps: tempo/rir/rpe are targets (prescribed but not logged); reps is not (it is both)', () => {
        const fields = targetOnlyFieldsFor({ tracking_type: 'sets_reps', tracked_metrics: ['tempo', 'rir', 'rpe'] });
        expect(fields.sort()).toEqual(['rest_seconds', 'rir', 'rpe', 'tempo'].sort());
    });

    test('time_based: rest_seconds is the only target-only field (prescribed base, but not a loggable metric)', () => {
        expect(targetOnlyFieldsFor({ tracking_type: 'time_based', tracked_metrics: ['duration_seconds', 'distance_km'] }))
            .toEqual(['rest_seconds']);
    });
});

File: mobile/test/unit/exercise_tracking_types_test.dart
import 'package:fitforce_x/shared/utils/exercise_tracking_types.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('categoryOf', () {
    test('returns the given category when valid', () {
      expect(categoryOf('time_based'), 'time_based');
    });
    test('falls back to sets_reps for null/unknown', () {
      expect(categoryOf(null), 'sets_reps');
      expect(categoryOf('bogus'), 'sets_reps');
    });
  });

  group('trackedMetricsOf', () {
    test('filters to only the category\'s selectable metrics, in order', () {
      expect(
        trackedMetricsOf('sets_reps', ['rir', 'rpe', 'distance_km']),
        ['rir', 'rpe'],
      );
    });
    test('null trackedMetrics yields an empty list', () {
      expect(trackedMetricsOf('sets_reps', null), isEmpty);
    });
  });

  group('prescribedFieldsFor — sets_reps', () {
    test('always includes reps + rest_seconds even with no metrics', () {
      expect(prescribedFieldsFor('sets_reps', []),
          ['reps', 'rest_seconds']);
    });
    test('adds selected metrics in canonical order', () {
      expect(
        prescribedFieldsFor('sets_reps', ['rpe', 'tempo']),
        ['reps', 'rest_seconds', 'tempo', 'rpe'],
      );
    });
  });

  group('prescribedFieldsFor — time_based', () {
    test('always includes rest_seconds even with no metrics', () {
      expect(prescribedFieldsFor('time_based', []), ['rest_seconds']);
    });
    test('adds selected metrics in canonical order', () {
      expect(
        prescribedFieldsFor('time_based', ['speed_kmh', 'duration_seconds']),
        ['duration_seconds', 'speed_kmh', 'rest_seconds'],
      );
    });
  });

  group('loggedFieldsFor', () {
    test('sets_reps is always just weight + reps, metrics never loggable', () {
      expect(loggedFieldsFor('sets_reps', ['tempo', 'rir', 'rpe']),
          ['weight', 'reps']);
    });
    test('time_based has no base fields, only selected metrics', () {
      expect(loggedFieldsFor('time_based', []), isEmpty);
      expect(
        loggedFieldsFor('time_based', ['distance_km', 'duration_seconds']),
        ['duration_seconds', 'distance_km'],
      );
    });
  });

  group('targetOnlyFieldsFor', () {
    test('sets_reps: prescribed but not loggable (tempo/rir/rpe), never reps', () {
      expect(
        targetOnlyFieldsFor('sets_reps', ['tempo', 'rir', 'rpe']),
        ['rest_seconds', 'tempo', 'rir', 'rpe'],
      );
    });
    test('time_based: selected metrics are loggable, so no targets beyond rest', () {
      expect(
        targetOnlyFieldsFor('time_based', ['distance_km', 'speed_kmh']),
        ['rest_seconds'],
      );
    });
  });
}
